### ISSUE TITLE
stream_socket_client() returns a file stream if transport not specified

### ISSUE THREAD

**Issue by Krapulat** [facebook/hhvm#5773]
PHPMailer doesn't work well with HHVM.  Using the lastest version: https://github.com/PHPMailer/PHPMailer and using SMPT AUTH with TLS.

The error is:
Fatal error: Unexpected object type File. in /var/www/domain/class.smtp.php on line 344

The line is this:
stream_socket_enable_crypto($this->smtp_conn,true,STREAM_CRYPTO_METHOD_TLS_CLIENT)

With PHP-FPM it works fine.

**Comment by lots0logs** [comment]
+1 I just ran into this issue today :confused:

**Comment by 64616E69656C** [comment]
to follow up on https://github.com/facebook/hhvm/issues/1647#issuecomment-123283860, here's a cut-down version of my PHPMail implementation:
https://gist.github.com/64616E69656C/47a9e4a1c95b745b9b8c

**Comment by fredemmott** [comment]
PHPMailer is depending on undocumented behavior - repro:

```Hack
<?php
// This is incorrect, though PHP5 allows it.
// per documentation, the format must be transport://target:port, not just
// target:port
$x = stream_socket_client('example.com:443');
$y = stream_socket_enable_crypto($x, true, STREAM_CRYPTO_METHOD_TLS_CLIENT);
var_dump($y);
```

Workaround: I suspect specifying 'tcp://localhost' as your host will work.

HHVM should still be modified to match PHP5's behavior.

**Comment by fredemmott** [comment]
Putting in a fix for stream_socket_client - confirmed it works with stream_socket_enable_crypto, but not set up a mailserver/phpmailer. Would appreciate it if someone can confirm that phpmailer works with this change:

```
diff --git a/hphp/runtime/base/ssl-socket.cpp b/hphp/runtime/base/ssl-socket.cpp
index 63eff03..1091426 100644
--- a/hphp/runtime/base/ssl-socket.cpp
+++ b/hphp/runtime/base/ssl-socket.cpp
@@ -372,7 +372,10 @@ req::ptr<SSLSocket> SSLSocket::Create(
     method = CryptoMethod::ClientSSLv3;
   } else if (scheme == "tls") {
     method = CryptoMethod::ClientTLS;
-  } else if (scheme == "tcp") {
+  } else if (
+    scheme == "tcp"
+    || (scheme.empty() && (domain == AF_INET || domain == AF_INET6))
+  ) {
     method = CryptoMethod::NoCrypto;
   } else {
     return nullptr;
```

**Comment by 64616E69656C** [comment]
As for the workaround, tcp://localhost (i did try 127.0.0.1 w/o tcp:// since connecting to localhost sometimes uses sockets) w/ or w/o port does not work:
```Fatal error: Uncaught exception 'phpmailerException' with message 'SMTP connect() failed.'```

tls:// does, but with the original error.

I can't test the patch myself though since i'm on deb releases and building hhvm from source on my dev machine is still on my todo, but I don't have enough coffee (well, and spare time) ATM :-/
However, if I understand the patch correctly, this disables STARTTLS when either scheme tcp or no scheme is given? Don't get it, because local relay w/o STARTTLS and w/o AUTH on tcp://localhost:25 was and is already working.

**Comment by fredemmott** [comment]
- if tcp:// is specified, an SSLSocket is initialized with no crypto
- if stream_socket_enable_crypto is called, nocrypto is changed (usually to tlsclient)
- without this change, if no transport is specified, an SSLSocket is *not* initialized, so there's nothing for stream_socket_enable_crypto() to change the crypto method on.

**Comment by Krapulat** [comment]
Adding "tcp://" to host, it shows this error:

    Warning: SSL operation failed with code 1. OpenSSL Error messages:\nerror:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed in /var/www/domain.com/wp-includes/class-smtp.php on line 262

The line is the same as before:
    stream_socket_enable_crypto($this->smtp_conn,true,STREAM_CRYPTO_METHOD_TLS_CLIENT)

**Comment by fredemmott** [comment]
Is the SSL certificate for your email server signed by a trusted CA?

**Comment by Krapulat** [comment]
No.

I've just tried it with one server signed by a trusted CA and IT WORKS.

So is it no possible to send mail by a server with a self signed SSL key?

**Comment by fredemmott** [comment]
See http://php.net/manual/en/context.ssl.php

If you're absolutely sure that you do not need to verify the certificate (eg if you're using localhost, not talking over the internet), you want  ['ssl' => ['verify_peer' => false]] in the stream context - it looks like for PHPMailer that can just go in the '$options' parameter

**Comment by 64616E69656C** [comment]
@fredemmott ugh, got it. Had to look at the whole file, because context matters - at 11pm.
I asked this one before - is there any chance for a deb-src, in order to apply such small fixes to a stable release w/ quilt? :-)

**Comment by fredemmott** [comment]
we don't built our packages in the usual manner - we don't have a deb-src to share :( This fix will hopefully be in 3.8.1
